46.md

```diff
@@ -1,98 +1,162 @@
+
+In addition, it's important that clients validate that the pubkey of the announced remote signer matches the pubkey of the `_` entry in the `/.well-known/nostr.json` file of the remote signer's announced domain.
+
+Clients that allow users to create new accounts should also consider validating the availability of a given username in the namespace of remote signer's domain by checking the `/.well-known/nostr.json` file for existing usernames. Clients can then show users feedback in the UI before sending a `create_account` event to the remote signer and receiving an error in return. Ideally, remote signers would also respond with understandable error messages if a client tries to create an account with an existing username.
+
+## The flow
+
+### Client-side
 
-## Event payloads
+1. Clients need several things to start: a local keypair and the details of the pubkey and remote signer the user would like to use.
+   - The local keypair is roughly disposable. E.g. - it's ok to store it locally (ideally encrypted) for some period of time. Once authenticated, remote signers can choose to trust this local keypair for a period of time. This local keypair can be deleted/rotated at will by the client and will only trigger a new auth challenge or permissions on the remote signer.
+   - The remote pubkey is the pubkey that the user is trying to sign in / sign as. This can be fetched with an NIP-05 or entered directly by the user (e.g. by passing a `bunker://` token).
+2. Clients use the local keypair for encrypting request events that they send to remote signers.
 
-Event payloads are [NIP-04](04.md)-encrypted JSON blobs that look like JSONRPC messages (their format is specified inside the `.content` of the event formats below).
+### Remote signer-side
 
-Events sent by the client to the remote signer have the following format:
+1. Remote signers manage the private key of the user that will be used for signing. There are many different remote signers out there (e.g. Nsecbunker, Amber, etc.) that operate in different ways and have different features.
+2. Remote signers wait for request events and then act.
 
-```js
+### Example flow for signing an event
+
+Coming soon...
+
+### Example Oauth-like flow to create a new user account with Nsecbunker
+
+Coming soon...
+
+## Request Events `kind: 24133`
+
+```json
 {
-  "pubkey": "<client-key-hex>"
-  "kind": 24133,
-  "tags": [
-    ["p", "<signer-key-hex>"]
-  ],
-  "content": "nip04_encrypted_json({id: <random-string>, method: <see-below>, params: [array_of_strings]})",
-  ...
+    "id": <id>,
+    "kind": 24133,
+    "pubkey": <local_keypair_pubkey>,
+    "content": <nip04(<request>)>,
+    "tags": [["p", <remote pubkey>]],
```

Comment:
Unless it's a create_account/connect_with_username, in which case it's the signer's pubkey. This is my very least favorite part of the spec.

Reply:
What about adding somewhere: `p tag MAY be the signer's pubkey, or if known, the user's pubkey. Using a user's pubkey makes it possible to connect to any signer that custodies the user's key, but if a specific signer has already been selected, it may be preferable to continue p-tagging the signer instead.` This could actually fix all my major objections to the current form of the protocol.

Reply:
I don't think that you could _just_ tag the signer in the `p` tag. There needs to be at least one `p` tag with the remote pubkey for most commands. Otherwise signers don't know which pubkey to use to decrypt the payload. 

I understand what you mean with `create_account` using the signer's pubkey and that being confusing. I can definitely add a comment about that.

Reply:
also @staab I believe that you are still using the remote user pubkey in the `p` tag for the connect method.

Reply:
The entire concept of the `remote signer pubkey` is ONLY useful when you need to perform tasks on the bunker (like `create_account`) itself.

the `p` tag should always be either the local user pubkey or remote user pubkey (i.e. the counterparty of the communication)

If we moved the `create_account` into a separate part of the NIP this might become way simpler since we don't even need to mention until that part that the signer itself can have a pubkey.